[Monolog] Added ElasticsearchLogstashHandler

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

This PR was initially [submitted on Monolog](https://github.com/Seldaek/monolog/pull/1334).
It has been refused , but Jordi suggested to add it to the symfony bridge. So here we go :)

---

ATM, there are few options to push log to Elastic Stack in order to play with them:

* install logstash and use a Gelf handler. It works but you have to install logstash and configure it. Not an easy task. More over, it need an extra PHP package
* use the ES handler: It does not play well with context and extra: Kibana is not able to filter on nested object. And this handler is tightly coupled to the ElasticaFormatter formater. More over, it need an extra PHP package
* use something to parse file logs. This is really a bad idea since it involves a parsing... More over a daemon is needed to do that (file beat / logstash / you name it)

This is why I'm introducing a new Handler.

* There is not need to install anything (expect ES, of course)
* It play very well with Kibana, as it uses the Logstash format
* It requires symfony/http-client, but in a modern PHP application (SF 4.3) this dependency is already present 
* It slow down a bit the application since it trigger an HTTP request for each logs. But symfony/http-client is non-blocking. If you want to use it in production, I recommend to wrap this handler in a buffer handler or a cross-finger handle to have only one HTTP call.

---

Some performance consideration en a prod env with a buffer handler + this one

* with push to ES: https://blackfire.io/profiles/f94ccf35-9f9d-4df1-bfc5-7fa75a535628/graph
* with push to ES commented: https://blackfire.io/profiles/6b66bc18-6b90-4341-963f-797f7a7a689c/graph

As you can see, as requests are made synchronously, there is no penalty on `AppKernel::Handler()` :heart_eyes:! But the PHP worker has more work to do, and it's busy much more time (about X2)

I explained everything in the PHP Doc Block

---

This is what you can expect **out of the box**
![image](https://user-images.githubusercontent.com/408368/59916122-9b7b7580-941e-11e9-9a22-f56bc1d1a288.png)

lyrixx committed:
[Monolog] Added ElasticsearchLogstashHandler

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,77 @@
+    private $client;
+    private $responses;
+
+    public function __construct($endpoint = 'http://elasticsearch:9200', $index = 'monolog', HttpClientInterface $client = null, $level = Logger::DEBUG, $bubble = true)
```
missing `string` types I suppose?

lyrixx replied:
Yes, indeed. I removed them for monolog 1.x

nicolas-grekas commented:
> Note: This handler is not aimed to be used in productio

I'd suggest telling this in the docblock +  explain how to do it in prod

lyrixx commented:
> I'd suggest telling this in the docblock + explain how to do it in prod

Actually, I could work a bit on it to make it available on production:
It it's used in combinaison with a buffer handler, or a finger cross handler, all logs could be sent by batch, and so only one HTTP call.

derrabus commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,85 @@
+        $this->sendToElasticsearch([$record]);
+    }
+
+    protected function getDefaultFormatter()
```
Same here, let's add a `FormatterInterface` return type.

https://github.com/Seldaek/monolog/blob/e07c948042045c9acff160648db45e40f75b1c58/src/Monolog/Handler/FormattableHandlerTrait.php#L59

ro0NL commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,85 @@
+
+    public function __construct(string $endpoint = 'http://elasticsearch:9200', string $index = 'monolog', HttpClientInterface $client = null, int $level = Logger::DEBUG, bool $bubble = true)
+    {
+        if (!class_exists(HttpClientInterface::class)) {
```
`interface_exists` (but tests will tell :))

fabpot commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,85 @@
+    public function __construct(string $endpoint = 'http://elasticsearch:9200', string $index = 'monolog', HttpClientInterface $client = null, int $level = Logger::DEBUG, bool $bubble = true)
+    {
+        if (!class_exists(HttpClientInterface::class)) {
+            throw new \LogicException(sprintf('%s handler needs symfony/http-client, please run `composer require symfony/http-client`.', __CLASS__));
```
The %s handler

fabpot reviewed (changes requested):
Your suggestion for a prod mode makes sense to me.

lyrixx commented:
* I updated the PHPdoc. 
* I added blackfire profile to the PR description
* I enhanced the main dockbloc
* I added an entry to the readme
* I changed the default host to 127.0.0.1:9200
* I added a timeout to 1 sec

lyrixx commented:
@nicolas-grekas How can I disable **all exceptions** in the http-client? Or at least catch them all? (But I think this is not do-able since I use the __destruct feature)

lyrixx commented:
Hmm, I managed to do something. Could you look at the last commit? Thanks

lyrixx commented:
I rebased this PR. It's now ready and green :green_heart:

noniagriconomie reviewed (commented):
minor doc review :)

edit : do you plan to add real doc, or a blog post to explain the config for dev/vs prod ?
i think it is a must have, to not degrade perf on prod :)

lyrixx commented:
> edit : do you plan to add real doc, or a blog post to explain the config for dev/vs prod ?
> i think it is a must have, to not degrade perf on prod :)

Not really. There is already many blog post on internet about symfony + ELK

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+    private function wait(?float $timeout, bool $errorOnTimeout)
+    {
+        $e = null;
+        $remainingTimeout = $timeout;
```
$remainingTimeout is always falsy so we can remove it, isn't it?

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+        }
+
+        if ($e) {
+            throw $e;
```
can be replaced by `error_log(sprintf("Could not push logs to elasticsearch:\n%s", (string) $e));` and the try/catch be removed above

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+/**
+ * Push logs directly to Elasticsearch and format them according to Logstash specification.
+ *
+ * This handler dials directly with the HTTP interface of elasticsearch. This
```
Elasticsearch (same on the line below)

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+ * answer. Even if all HTTP calls are done asynchronously.
+ *
+ * In a development environment, it's fine to keep the default configuration:
+ * For each log, an HTTP request will be made to push the log to Elasticsearch.
```
s/For/for

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+ * In a development environment, it's fine to keep the default configuration:
+ * For each log, an HTTP request will be made to push the log to Elasticsearch.
+ *
+ * But in a production environment, it's highly recommended to wrap this handler
```
s/But in/In

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+ * For each log, an HTTP request will be made to push the log to Elasticsearch.
+ *
+ * But in a production environment, it's highly recommended to wrap this handler
+ * in an handler with buffering capability (like the FingersCrossedHandler, or
```
s/an handler/a handler
/capability/capabilities

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+    public function __construct(string $endpoint = 'http://127.0.0.1:9200', string $index = 'monolog', HttpClientInterface $client = null, int $level = Logger::DEBUG, bool $bubble = true)
+    {
+        if (!interface_exists(HttpClientInterface::class)) {
+            throw new \LogicException(sprintf('The %s handler needs symfony/http-client, please run `composer require symfony/http-client`.', __CLASS__));
```
The %s handler needs an HTTP client. Try running "composer require symfony/http-client".

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+
+        $this->sendToElasticsearch([$record]);
+
+        return false === $this->bubble;
```
`!$this->bubble`?

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+            return $this->isHandling($record);
+        });
+
+        if (!$records) {
```
```php
if ($records) {
    $this->sendToElasticsearch($records);
}
```

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,169 @@
+
+    public function handleBatch(array $records): void
+    {
+        $records = array_filter($records, function (array $record) {
```
`$records = array_filter($records, [$this, 'isHandling']);`

lyrixx commented:
I addressed your issues

nicolas-grekas commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,149 @@
+        $this->wait(null, true);
+    }
+
+    private function wait(?float $timeout, bool $errorOnTimeout)
```
maybe change the signature to `wait(bool $blocking)`?

lyrixx replied:
Why `blocking`? It should be `errorOnTimeout`, isn't ?

lyrixx replied:
Fixed :+1:

nicolas-grekas reviewed (approved):
Nice :)

noniagriconomie commented:
@lyrixx 

> Not really. There is already many blog post on internet about symfony + ELK

my comment was more for configuring **this new handler** in the proper way inside a symfony **dev|prod** stack :)

lyrixx commented:
> my comment was more for configuring **this new handler** in the proper way inside a symfony **dev|prod** stack :)

This will be done in the monolog-bundle :)

Taluu commented on src/Symfony/Bridge/Monolog/Handler/ElasticsearchLogstashHandler.php:
```diff
@@ -0,0 +1,144 @@
+
+        $body = '';
+        foreach ($records as $record) {
+            if ($this->processors) {
```
can't remember how the processors property is registered, but if it's always an array you can skip the `if`

lyrixx replied:
Good catch! Thanks. Fixed

B-Galati commented:
Do you plan to add some doc ?

B-Galati commented on src/Symfony/Bridge/Monolog/Tests/Handler/ElasticsearchLogstashHandlerTest.php:
```diff
@@ -0,0 +1,119 @@
+{
+    protected function getDefaultFormatter(): FormatterInterface
+    {
+        return new LogstashFormatter('application', 'my hostname', null, 'ctxt_', LogstashFormatter::V1);
```
The DefaultFormatter that comes with the handler is not tested though.

Perhaps you want to consider adding a simple test with it.

fabpot commented:
Thank you @lyrixx.

nicolas-grekas commented:
@lyrixx could you please have a look at failures on branch master? It looks like the new class is not compatible with Monolog 2. Thanks.

lyrixx commented:
:+1: I will look at it. Thanks

lyrixx commented:
https://github.com/symfony/symfony/pull/33241

DavidBadura commented:
thanks @lyrixx for this contribution :) 

I tried it and found the problem that the logging recrudes. The handler uses the HttpClient, which logs every request (`POST http://127.0.0.1:9200/_bulk`). I had to exclude the channel `http_client`, otherwise I hit the memory limit.

```
monolog:
    handlers:
        es:
            type: service
            id: Symfony\Bridge\Monolog\Handler\ElasticsearchLogstashHandler
            channels: ["!http_client"]
```

Do you have any idea how to still log the http client requests without landing in a recrusion?
I think, the problem should also be documented.

DavidBadura commented:
I bypassed this problem by not using the global client:

```yaml
    Symfony\Bridge\Monolog\Handler\ElasticsearchLogstashHandler:
        arguments:
            $client: null
```

lyrixx commented:
Hello. I did not anticipate this :joy: 
I'm working on a fix.

lyrixx commented:
Hmm, I don"t see how Symfony could fix this issue since it's "normal" to have this behavior.
So instead I submitted a PR to the doc repo https://github.com/symfony/symfony-docs/pull/12642
and I fixed my article https://jolicode.com/blog/how-to-visualize-symfony-logs-in-dev-with-elasticsearch-and-kibana

Thanks for the feedback :+1:
